x/epochstorage/README.md

```diff
@@ -0,0 +1,172 @@
+# `x/epochstorage`
+
+## Abstract
+
+This document specifies the epochstorage module of Lava Protocol.
+
+Lava blocks are divided into bulk of blocks called epochs. epoch length is defines as a parameter of the module.
+The module keeps track of the epochs in memory, epochs that needs to be deleted and anchering data to each epoch, mainly providers stakes.
+note that the module will make sure that any changes will take effect only in the next epoch.
+
+## Contents
+* [Concepts](#concepts)
+  * [Epoch](#epoch)
+    * [EpochDetails](#epochdetails)
+  * [FixatedParams](#fixatedparams)
+  * [StakeStorage](#stakeStorage)
+  * [StakeEntry](#stakeentry)
+* [Parameters](#parameters)
+* [Queries](#queries)
+* [Transactions](#transactions)
+* [Proposals](#proposals)
+* [Events](#events)
+
+## Concepts
+
+### Epoch
+
+Epoch is defined as a batch of blocks, for each epoch all the data that affects pairing stays the same. for example: all the providers and their data are the same for all the blocks of the epochs and can be changed only in the next epoch.
+
+#### EpochDetails
+
+This object shows the details of the current epoch, the last epoch in memory and the last deleted epochs.
+
+```go
+type EpochDetails struct {
+	StartBlock    uint64   // start of the current epoch
+	EarliestStart uint64   // earliest epoch that is saved on memory
+	DeletedEpochs []uint64 // list of deleted epochs (updates each epoch)
+}
+```
+
+### FixatedParams
+
+Some params on the chain can affect the pairing list results, since pairing list needs to be the same throughout the epoch, the chain saves the needed params per epoch. this is done with fixated params. each module with a param that affects the pairing list needs to register the param as a fixated param using the following method of the keeper:
+
+```go
+AddFixationRegistry(fixationKey string, getParamFunction func(sdk.Context) any) 
+```
+
+by doing so the module will know to track [param change proposals](../spec/proposal_handler.go) and save any changes of the relevant param as a fixated param on the store.
+fixated param is in charge of tracking param changes, save their different versions if they are registered and delete old versions when they are not needed anymore (determined by EpochsToSave parameter).
+
+a fixation parameter is saved on store as:
+
+```go
+
+type FixatedParams struct {
+	Index         string // the index of the param
+	Parameter     []byte // serialized data of the param
+	FixationBlock uint64 // the block at which the param has changed
+}
+```
+
+this is done in the [begin block of the module](keeper/fixated_params.go)
```

Comment:
this -> This
begin block -> BeginBlock method

Reply:
done